File: java-sdk/radar-schemas-tools/src/main/java/org/radarbase/schema/tools/ListCommand.java
package org.radarbase.schema.tools;

import java.util.stream.Collectors;
import java.util.stream.Stream;
import net.sourceforge.argparse4j.impl.Arguments;
import net.sourceforge.argparse4j.inf.ArgumentParser;
import net.sourceforge.argparse4j.inf.Namespace;

public class ListCommand implements SubCommand {

    @Override
    public String getName() {
        return "list";
    }

    @Override
    public int execute(Namespace options, CommandLineApp app) {
        Stream<String> out;
        if (options.getBoolean("raw")) {
            out = app.getRawTopics();
        } else if (options.getBoolean("stream")) {
            out = app.getResultsCacheTopics();
        } else if (options.getBoolean("quiet")) {
            out = app.getTopicsToCreate();
        } else {
            out = app.getTopicsVerbose(true, options.getString("match"));
        }
        System.out.println(out.sorted().distinct()
                .collect(Collectors.joining("\n")));
        return 0;
    }

    @Override
    public void addParser(ArgumentParser parser) {
        parser.description("list topics and schemas");
        parser.addArgument("-r", "--raw")
                .help("list raw input topics")
                .action(Arguments.storeTrue());
        parser.addArgument("-q", "--quiet")
                .help("only print the requested topics")
                .action(Arguments.storeTrue());
        parser.addArgument("-m", "--match")
                .help("only print the requested topics");
        parser.addArgument("-S", "--stream")
                .help("list the output topics of Kafka Streams")
                .action(Arguments.storeTrue());
        SubCommand.addRootArgument(parser);
    }
}

File: java-sdk/radar-schemas-tools/src/main/java/org/radarbase/schema/tools/ListCommand.kt
package org.radarbase.schema.tools

import net.sourceforge.argparse4j.impl.Arguments
import net.sourceforge.argparse4j.inf.ArgumentParser
import net.sourceforge.argparse4j.inf.Namespace
import org.radarbase.schema.tools.SubCommand.Companion.addRootArgument
import java.util.stream.Collectors
import java.util.stream.Stream

class ListCommand : SubCommand {
    override val name: String = "list"

    override suspend fun execute(options: Namespace, app: CommandLineApp): Int {
        val out: Stream<String> = when {
            options.getBoolean("raw") -> app.rawTopics
            options.getBoolean("stream") -> app.resultsCacheTopics
            options.getBoolean("quiet") -> app.topicsToCreate
            else -> app.getTopicsVerbose(true, options.getString("match"))
        }
        println(
            out
                .sorted()
                .distinct()
                .collect(Collectors.joining("\n")),
        )
        return 0
    }

    override fun addParser(parser: ArgumentParser) {
        parser.apply {
            description("list topics and schemas")
            addArgument("-r", "--raw")
                .help("list raw input topics")
                .action(Arguments.storeTrue())
            addArgument("-q", "--quiet")
                .help("only print the requested topics")
                .action(Arguments.storeTrue())
            addArgument("-m", "--match")
                .help("only print the requested topics")
            addArgument("-S", "--stream")
                .help("list the output topics of Kafka Streams")
                .action(Arguments.storeTrue())
            addRootArgument()
        }
    }
}
